cond_variable_instead_sleep_for.patch

```diff
@@ -0,0 +1,46 @@
+--- worker.hpp	Thu Mar 22 23:20:12 2018
++++ worker.hpp	Sat Mar 24 22:28:13 2018
+@@ -2,6 +2,8 @@
+ 
+ #include <atomic>
+ #include <thread>
++#include <condition_variable>
++#include <mutex>
+ 
+ namespace tp
+ {
+@@ -78,6 +80,8 @@
+     Queue<Task> m_queue;
+     std::atomic<bool> m_running_flag;
+     std::thread m_thread;
++    std::mutex m_conditional_mutex;
++    std::condition_variable m_conditional_lock;
+ };
+ 
+ 
+@@ -121,6 +125,7 @@
+ inline void Worker<Task, Queue>::stop()
+ {
+     m_running_flag.store(false, std::memory_order_relaxed);
++    m_conditional_lock.notify_all();
+     m_thread.join();
+ }
+ 
+@@ -140,6 +145,7 @@
+ template <typename Handler>
+ inline bool Worker<Task, Queue>::post(Handler&& handler)
+ {
++    m_conditional_lock.notify_one();
```

Comment:
I think you should do the notification after you have added the job to the queue otherwise the worker might try to pull the job too soon

Reply:
Мммммммм, in theory may be. In practice....hardly. Also, how you imagine such implementation? I see only one way - additional local variable requires, so more assembler, slower execution... No, not so good idea.

Reply:
If you're going to worry about the cost of that it might be more interesting to look at the call to `notify_all` you have for every posted job. Compared to that cost saving the bool result is essentially free.

Reply:
> 
> 
> If you're going to worry about the cost of that it might be more interesting to look at the call to `notify_one` you have for every posted job. Compared to that cost saving the bool result is essentially free.

Take a look on my own fork of this library. Done long time ago. :) This PR is years old :)

Anyway, you can make you own fork, perform benchmark and compare. :)